From b581e4a00fcab210172589c878b13892b26f43ed Mon Sep 17 00:00:00 2001
From: EKR <ekr@rtfm.com>
Date: Wed, 11 Jan 2017 14:36:35 -0800
Subject: [PATCH] Force short headers

---
 lib/ssl/manifest.mn | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lib/ssl/manifest.mn b/lib/ssl/manifest.mn
index e7564ed..5994725 100644
--- a/lib/ssl/manifest.mn
+++ b/lib/ssl/manifest.mn
@@ -55,3 +55,5 @@ LIBRARY_VERSION = 3
 
 # This part of the code, including all sub-dirs, can be optimized for size
 export ALLOW_OPT_CODE_SIZE = 1
+
+DEFINES += -DNSS_ENABLE_TLS13_SHORT_HEADERS
\ No newline at end of file
